keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,327 @@
+default behavior as if the ImageDecryptSecret feature is disabled.
+
+#### Examples
+
+Suppose we operate a cluster that lets users create a secret of type `ImageDecryptSecret`. 
+
+For the kubectl command line, we propose that the user create a secret of type `image-decrypt` and give it a name. One or multiple private keys may then be stored under this secret.
+
+```bash
+kubectl create secret image-decrypt <secret name> --decrypt-secret=<private key>[:<password>] [--decrypt-secret=<private key>[:<password>]]
+```
+`private key` - base64 representation of key
+
+`password` (optional) - password in *cleartext* if the given private key is protected with the password
+
+For example,
+```bash
+kubectl create secret image-decrypt keysecret --decrypt-secret="base64OfPrivateKey":"password"
+```
+See it in action in the [demo](https://youtu.be/S3FK4y5McOk?t=19)
+
+A variant for passing the ImageEncrypSecret is the following:
+
+```bash
+kubectl create secret image-decrypt <secret name> --from-file=<path to file>:[:<password>] 
+```
+
+`path to file` - The path of the private key in binary format
+
+`password` (optional) - password in *cleartext* if the given private key is protected with the password
+For example,
+```bash
+kubectl create secret image-decrypt keysecret --from-file="/path/to/private_key":"password"
+```
+
+
+Then when a user creates a workload, they can choose the desired ImageDecryptSecret to use 
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  labels:
+    name: nginx
+spec:
+  containers:
+  - name: nginx
+    image: localhost:5000/nginx:enc
+    imageDecryptSecrets:
+    - name: <secret name>
+    imagePullPolicy: Always
+    ports:
+    - containerPort: 80
+
+```
+
+ImageDecryptSecrets can be added to the `service account` by,
+
+```bash
+kubectl patch serviceaccount <account name> -p '{"imageDecryptSecrets":[{"name":<secret name>, "images":[<image name>}]}'
+```
+or while creating a secret account,
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  creationTimestamp: 2015-08-07T22:02:39Z
+  name: default
+  namespace: default
+  selfLink: /api/v1/namespaces/default/serviceaccounts/default
+  uid: 052fb0f4-3d50-11e5-b066-42010af0d7b6
+secrets:
+- name: default-token-uudge
+imagePullSecrets:
+- name: myregistrykey
+imageDecryptSecrets:
+- name: <secret name>
+- image: [<image name>]
+```
+
+For example,
+```bash
+kubectl patch serviceaccount default -p '{"imageDecryptSecrets":[{"name":"keysecret", "images":["localhost:5000/nginx:enc"]}]}'
+```
+
+#### Image Handler
+
+The privake keys are extracted from `ImageDecryptSecret` and passed to the CRI as by bundling them in `DecryptParams` as a part of `PullImageRequest`:
+
+
+```protobuf
+// DecryptParams is used to send decryption parameters to the image service of the CRI
+message DecryptParams {
+    repeated string privateKeyPasswds = 1;
+}
+```
+```protobuf
+message PullImageRequest {
```

Comment:
The container creation api change described in https://github.com/kubernetes/enhancements/pull/1066#issuecomment-520703126 is not covered.

Reply:
Addressed it [here](https://github.com/kubernetes/enhancements/blob/edc643b6a55a2ed4cc0231447352bc3ef65136e7/keps/sig-node/20190517-image-decryption.md#Container-handler).